src/ripple/app/main/NodeStoreScheduler.cpp

```diff
@@ -48,16 +46,21 @@ void NodeStoreScheduler::onChildrenStopped ()
 void NodeStoreScheduler::scheduleTask (NodeStore::Task& task)
 {
     ++m_taskCount;
-    m_jobQueue->addJob (
+    if (!m_jobQueue->addJob (
         jtWRITE,
         "NodeObject::store",
-        [this, &task] (Job&) { doTask(task); });
+        [this, &task] (Job&) { doTask(task); }))
+    {
+        // Job not added, presumably because we're shutting down.
+        // Recover by executing the task synchronously.
+        doTask (task);
+    }
 }
 
 void NodeStoreScheduler::doTask (NodeStore::Task& task)
 {
     task.performScheduledTask ();
-    if ((--m_taskCount == 0) && isStopping())
+    if ((--m_taskCount == 0) && isStopping() && areChildrenStopped())
```

Comment:
That diagram is extremely useful.  I'd consider adding it to the code somewhere, even if it might get stale.

I'll need to study this section a bit more as well and am not advocating a change yet.